### PR TITLE
ModularUI Part 1

### PR DESCRIPTION
**Pull request by brachy84**
This will be the start of implementing ModularUI into GTCEu.
This PR will contain all covers and non generic MTEs

The filter system was rewritten in this PR. Please test compatibility with older versions:
- set up some covers on **master** branch
- configure them (transfer amount, filters, etc...)
- switch to this branch and reload world
- everything still the same?

- [x] Conveyor
- [x] Robot Arm
- [x] Pump
- [x] Regulator
- [x] (Advanced) Item Voiding
- [x] (Advanced) Fluid Voiding
- [x] Filter Cover
- [x] Digital Interface
- [x] Ender Fluid 
- [x] Machine controller
- [ ] Workbench (Block and Cover) (Fuck)
- [x] Crates (invtweaks compat only works in non dev as far as i tested)
- [x] Integrated Circuit

### CODE REVIEW

**Review by ALongStringOfNumbers — COMMENTED** [review]
Some requests/suggestions. These will be a mix of things for the implementation in CEu and for Modular UI itself.

1. For the Filter Settings popup window, after the window is opened, the button on the original UI still says "Open Settings". Would it be possible that when the Filter Configuration window is open that this button changes to "Close Settings" and then does the same thing as the `x` in the Filter Configuration window?

2. This one is a bit out there, but would it be possible for the Filter Configuration window to remember the spot that you moved it to, and then re-open in that place after closing the window and re-opening it?

3. When window animation opens and closes from the bottom, this only does the main window and not any secondary windows opened from the main window. EG when having a cover UI open and the Filter Configuration window open, only the cover UI closes to the bottom, and the Filter Configuration window just vanishes. How hard would it be to apply the window opening/closing effects to secondary windows as well as the main one?

4. When configuring the Fluid Filter from JEI ghost ingredients, would it be possible to also have filled Fluid Handlers that are in JEI be valid sources? EG a Bucket of Water can be used to put Water in the fluid filter.

5. When configuring a Fluid Filter with a Filled bucket, the bucket is tossed out of the player's inventory after successfully configuring the slot. This seems like a conflict with the click to fill/empty slot mechanic. (Also, what do you think about having this mechanic in Modular UI itself?)

6. If you open JEI with another window besides the main window open, Eg the Filter Configuration window, the secondary window resets its position to the original opening position.

7. The tooltip when hovering over the Filter Configuration window is drawn centered over the mouse, and will obstruct the slot that the player is interacting with. Seen here it covers the amount of fluid display in the slot. The tooltip should avoid the slot, and be drawn outside the slot's boundaries, so it does not obscure what is in the slot.



8. The Increment buttons in cover's UIs (Like the Advanced Item/Fluid Voiding cover in Void Overflow mode) could use tooltips to tell the player about the key combinations that can be used to change the amount incrememented.

9. When placing a filter as a cover and configuring the filter, the passed items/fluids take stack size into account, while before the rework they did not. Is this change intentional?

10. The tooltips for the Normal/Inverted button on the machine controller are inversed. In normal mode it shows the Inverted tooltip and vice versa.

11. There is some weirdness with the sliding bar on the machine controller. If I click a location in the widget, the bar moves to that slot, but then if I do not release my mouse and drag my mouse away from where I clicked, the bar does not follow my mouse. Not sure if this is intended or not.

**Comment by brachy84** [comment]
I can not reproduce number 5

**Comment by ALongStringOfNumbers** [comment]
As desired, I tested some of the cover saving stuff between `master` and this branch, and I found some issues.

----
Fluid Voiding Cover

I set up an regular fluid voiding cover with a fluid filter on master, and configured several fluids in the filter. When I switched to this branch, the fluid configuration was missing from the filter.

----
Advanced Item Voiding Cover

The Advanced Item voiding cover matched all information after switching branches

However, in the PR, the advanced item voiding cover can only have 64 items specified in an item slot in the filter in overflow mode, while on master you can have more than 64. The more than 64 stems from the maximum transfer amount per voiding mode in `VoidingMode`, where the maximum is set to 1024 for Overflow mode

In addition, the tooltip that displays the different key combinations you can hold, or that you can scroll wheel does not show up in the item filter, even though the key combos work and scroll works.

----
Advanced Fluid Voiding Cover

Similar to the Fluid Voiding cover, the fluid filter information is not saved in the Advanced Fluid Voiding cover.

----
General Cover stuff

For things like the Robot Arm, where there are multiple modes that can be selected from in a button, it would be nice if there was a tooltip that appeared over the entry name showing all the possible modes for the button.
For example, in this image


The tooltip would appear when hovering over `Distribution:` and show Enhanced RR, RR, and First Insert, and the same for the other rows where there are buttons with multiple options.


----
Robot Arm


When Configuring the Robot arm (but this may also happen on other covers), the tooltip for incrementing obscures the box where the current number is displayed.


With the Robot Arm in Keep Exact mode, you can no longer specify above 64 items in the item filter, you are able to do 1024 on `master`

The Supply Exact mode on the robot arm does not seem to function any more.
I have a robot arm on a crate set to export mode with supply exact 3 and several stacks of items in the crate. It is pointing into an empty crate next to it, but nothing is being transferred.



----
Crate

I saw a weird syncing issue with the crate.

If you open the Crate's GUI, and then cheat in items from JEI, the player's inventory is not updated, so the cheated in items do not display in the players inventory until you close and reopen the crate.

This is only for the player's inventory, and not for the hotbar, which does sync correctly.

**Comment by Syrcan** [comment]
Here's my review. I'm going by what I can see in-game.

----
**Pump / Regulator**
The description of the Bucket mode to Round Robin mode:
image

----
**Robot Arm / Regulator**
Transfer Mode displays `/n` as text. The description is also the same for both covers, which means the Regulator gets a description about items:
image

----
**Item Filter + Conveyor**
The window for the Item Filter Cover (attached to Conveyor/Robot Arm) feels extremely awkward with the slots sticking out, any reason on why it had to be a window and not an additional like with the Fluid Filter on a Pump?
image

*On a smaller note* (probably for MUI, not CEu), moving windows require clicking the window once to be able to move it, and clicking again to release it. Is it not possible to do click & hold?

----
**Item & Fluid Filters**
The covers support setting up stack size and fluid amount despite being irrelevant until they are attached to a Robot Arm / Regulator.
The Whitelist/Blacklist button does not display any tooltip.

----
**Fluid Filter**
Description for mouse shortcuts on the Fluid Filter are not accurate.
image
Clicking behaviors I tried:
Left click removes 1 bucket (-1,000L)
Right click adds 1 bucket (+1,000L)
Shift + left click clears the fluid
Shift + right click is the same as regular right click (+1 bucket)
Middle click is the same as left click (-1 bucket or clear when shift is also pressed)

----
**Item Filter**
Ignore damage and Ignore NBT button tooltips display a lang key instead of the (probably missing) lang entry.

----
**Ender Fluid Link**
They are using the old GUI despite being checked out as being done in your original post.

----
**Voiding Covers**
This is more of a *request* here. So you can ignore it. I would personally like that they get an Enable/Disable button in the GUI, rather than having to having to activate them with Soft Mallet.

----
**Open/Close Animation**
This one is more of a debatable topic. I think it looks cool and is beneficial to have for Covers.
However I think it is detrimental to have that behavior for the Crate (and other future inventories to get converted to MUI). Here's my explanation:
The player is often subject to open inventories in quick successions to check items, and the animation makes things slower. It does not feel responsive to close the inventory, and to be locked for half a second in the closing animation, being unable to move or do any other action.
That's one part of the issue, the other part is that you can interact with items during the animation, which makes it weird. For the brief moment you are opening the inventory, the cursor may be hovering over "nothing", but you will still have the item tooltip.
**TL;DR:** Yes for the animation to Covers and other configurable GUIs. No for the animation to inventories.

----
**Crate**
This may also apply to other inventories in MUI.
In the attached screenshot below, you can see the selection box is over the crate item, however clicking will pick up the bucket.
This screenshot capture tool hides the cursor (sad), but it is right in between both inventory slots.
After comparing to how a chest works, it may seem that the cursor position to pick up the item is correct in the Crate, but not the highlighted slot.
image

----
**Crash**
I got a client crash trying to open an Item Filter settings on a Robot Arm:
[crash-2022-06-15_07.07.01-client.txt](https://github.com/GregTechCEu/GregTech/files/8905858/crash-2022-06-15_07.07.01-client.txt)
I do not know how to reproduce, the game was open for a while, it just happened seemingly randomly.

**Comment by brachy84** [comment]
I'll close this for now,  as don't want to work on this anymore. Maybe after the MUI rework is done I'll come back to this.
